### ISSUE TITLE
Dealing with pixels

### ISSUE THREAD

**Issue by zxdawn** [climate-processes/tobac#5]
Since many satellites data are saved by **pixels**, maybe we need to improve the function of calculating dxdy:
https://github.com/climate-processes/tobac/blob/69da64a59a36e73a32f4cdff09a86508d966f351/tobac/utils.py#L414-L442

Or we can **grid** the satellite first and then use `tobac`.

**Comment by mheikenfeld** [comment]
Yes, that is definitely a good point, this function it really geared towards limited area model output in a specific format as it is at the moment. I have worked on calculating an approximate grid spacing based on the lat/lon values of the pixels e.g. for satellite data, will dig out that code and put it in as a start!

**Comment by zxdawn** [comment]
That's great! I'll use gridded data to try the package first.
Thank you for your excellent work!

**Comment by mheikenfeld** [comment]
Thanks to you for spotting all these things, good to see people interested and involved!

**Comment by JuliaKukulies** [comment]
I have code available where i modified the function get_spacings in the way that grid spacing can be estimated from lat and lon values if no other information is provided. Since I see that the function get_spacings has not been updated yet, I could create a pull request, if this is of interest? :) 

Another remark is that no error message appears when projection_x_coord and projection_y_coord is not defined. Instead you get 

**UnboundLocalError:** local variable 'dxy' referenced before assignment

This should be quickly fixed by modifying the else clause to

**else:** 
    raise ValueError(''no information about grid spacing, need either input cube with projection_x_coord and projection_y_coord or keyword argument grid_spacing'')

**Comment by JuliaKukulies** [comment]
A suggestion for calculating dxdy based on lats and lons: 


`
def get_spacings(field_in,grid_spacing=None,time_spacing=None):
    import numpy as np
    from copy import deepcopy
    from math import cos, sin, asin, sqrt, radians
    # set horizontal grid spacing of input data
    # If cartesian x and y corrdinates are present, use these to determine dxy (vertical grid spacing used to transfer pixel distances to real distances):
    coord_names=[coord.name() for coord in  field_in.coords()]
    print('Available coordinates:    ', coord_names)

    if (('projection_x_coordinate' in coord_names and 'projection_y_coordinate' in coord_names) and  (grid_spacing is None)):
        x_coord=deepcopy(field_in.coord('projection_x_coordinate'))
        x_coord.convert_units('metre')
        dx=np.diff(field_in.coord('projection_y_coordinate')[0:2].points)[0]
        y_coord=deepcopy(field_in.coord('projection_y_coordinate'))
        y_coord.convert_units('metre')
        dy=np.diff(field_in.coord('projection_y_coordinate')[0:2].points)[0]
        dxy=0.5*(dx+dy)
    elif grid_spacing is not None:
        dxy=grid_spacing

    elif (('longitude' in coord_names and 'latitude' in coord_names) and  (grid_spacing is None)):
        # get min and max values of lats and lons 
        lat1= np.min(field_in.coord('latitude').points)
        lat2= np.max(field_in.coord('latitude').points)

        lon1= np.min(field_in.coord('longitude').points)
        lon2= np.max(field_in.coord('longitude').points)

        # convert decimal degrees to radians
        lon1, lat1, lon2, lat2 = map(radians, [lon1, lat1, lon2, lat2])
        # haversine formula, note that value is divided by nr of grid cells along lat and lon 
        dlon = (lon2 - lon1) / np.shape(field_in.data)[1]
        dlat = (lat2 - lat1)/ np.shape(field_in.data)[2]
        a = sin(dlat / 2) ** 2 + cos(lat1) * cos(lat2) * sin(dlon / 2) ** 2
        c = 2 * asin(sqrt(a))
        km = 6371 * c
        dxy= np.round(km * 1000)
    else:
        raise ValueError('no information about grid spacing, need either input cube with longitude and latitude, projection_x_coord and projection_y_coord or keyword argument grid_spacing')
    
    # set horizontal grid spacing of input data
    if (time_spacing is None):    
        # get time resolution of input data from first to steps of input cube:
        time_coord=field_in.coord('time')
        dt=(time_coord.units.num2date(time_coord.points[1])-time_coord.units.num2date(time_coord.points[0])).seconds
    elif (time_spacing is not None):
        # use value of time_spacing for dt:
        dt=time_spacing
    return dxy,dt
`

**Comment by JuliaKukulies** [comment]
A suggestion for calculating dxdy based on lats and lons:

`
def get_spacings(field_in,grid_spacing=None,time_spacing=None):
    import numpy as np
    from copy import deepcopy
    from math import cos, sin, asin, sqrt, radians
    # set horizontal grid spacing of input data
    # If cartesian x and y corrdinates are present, use these to determine dxy (vertical grid spacing used to transfer pixel distances to real distances):
    coord_names=[coord.name() for coord in  field_in.coords()]
    print('Available coordinates:    ', coord_names)

    if (('projection_x_coordinate' in coord_names and 'projection_y_coordinate' in coord_names) and  (grid_spacing is None)):
        x_coord=deepcopy(field_in.coord('projection_x_coordinate'))
        x_coord.convert_units('metre')
        dx=np.diff(field_in.coord('projection_y_coordinate')[0:2].points)[0]
        y_coord=deepcopy(field_in.coord('projection_y_coordinate'))
        y_coord.convert_units('metre')
        dy=np.diff(field_in.coord('projection_y_coordinate')[0:2].points)[0]
        dxy=0.5*(dx+dy)
    elif grid_spacing is not None:
        dxy=grid_spacing

    elif (('longitude' in coord_names and 'latitude' in coord_names) and  (grid_spacing is None)):
        # get min and max values of lats and lons 
        lat1= np.min(field_in.coord('latitude').points)
        lat2= np.max(field_in.coord('latitude').points)

        lon1= np.min(field_in.coord('longitude').points)
        lon2= np.max(field_in.coord('longitude').points)

        # convert decimal degrees to radians
        lon1, lat1, lon2, lat2 = map(radians, [lon1, lat1, lon2, lat2])
        # haversine formula, note that value is divided by nr of grid cells along lat and lon 
        dlon = (lon2 - lon1) / np.shape(field_in.data)[1]
        dlat = (lat2 - lat1)/ np.shape(field_in.data)[2]
        a = sin(dlat / 2) ** 2 + cos(lat1) * cos(lat2) * sin(dlon / 2) ** 2
        c = 2 * asin(sqrt(a))
        km = 6371 * c
        dxy= np.round(km * 1000)
    else:
        raise ValueError('no information about grid spacing, need either input cube with longitude and latitude, projection_x_coord and projection_y_coord or keyword argument grid_spacing')
    
    # set horizontal grid spacing of input data
    if (time_spacing is None):    
        # get time resolution of input data from first to steps of input cube:
        time_coord=field_in.coord('time')
        dt=(time_coord.units.num2date(time_coord.points[1])-time_coord.units.num2date(time_coord.points[0])).seconds
    elif (time_spacing is not None):
        # use value of time_spacing for dt:
        dt=time_spacing
    return dxy,dt
`

**Comment by JuliaKukulies** [comment]
A suggestion for calculating dxdy based on lats and lons: 

```python

def get_spacings(field_in,grid_spacing=None,time_spacing=None):
    import numpy as np
    from copy import deepcopy
    from math import cos, sin, asin, sqrt, radians
    # set horizontal grid spacing of input data
    # If cartesian x and y corrdinates are present, use these to determine dxy (vertical grid spacing used to transfer pixel distances to real distances):
    coord_names=[coord.name() for coord in  field_in.coords()]
    print('Available coordinates:    ', coord_names)

    if (('projection_x_coordinate' in coord_names and 'projection_y_coordinate' in coord_names) and  (grid_spacing is None)):
        x_coord=deepcopy(field_in.coord('projection_x_coordinate'))
        x_coord.convert_units('metre')
        dx=np.diff(field_in.coord('projection_y_coordinate')[0:2].points)[0]
        y_coord=deepcopy(field_in.coord('projection_y_coordinate'))
        y_coord.convert_units('metre')
        dy=np.diff(field_in.coord('projection_y_coordinate')[0:2].points)[0]
        dxy=0.5*(dx+dy)
    elif grid_spacing is not None:
        dxy=grid_spacing

    elif (('longitude' in coord_names and 'latitude' in coord_names) and  (grid_spacing is None)):
        # get min and max values of lats and lons 
        lat1= np.min(field_in.coord('latitude').points)
        lat2= np.max(field_in.coord('latitude').points)

        lon1= np.min(field_in.coord('longitude').points)
        lon2= np.max(field_in.coord('longitude').points)

        # convert decimal degrees to radians
        lon1, lat1, lon2, lat2 = map(radians, [lon1, lat1, lon2, lat2])
        # haversine formula, note that value is divided by nr of grid cells along lat and lon 
        dlon = (lon2 - lon1) / np.shape(field_in.data)[1]
        dlat = (lat2 - lat1)/ np.shape(field_in.data)[2]
        a = sin(dlat / 2) ** 2 + cos(lat1) * cos(lat2) * sin(dlon / 2) ** 2
        c = 2 * asin(sqrt(a))
        km = 6371 * c
        dxy= np.round(km * 1000)
    else:
        raise ValueError('no information about grid spacing, need either input cube with longitude and latitude, projection_x_coord and projection_y_coord or keyword argument grid_spacing')
    
    # set horizontal grid spacing of input data
    if (time_spacing is None):    
        # get time resolution of input data from first to steps of input cube:
        time_coord=field_in.coord('time')
        dt=(time_coord.units.num2date(time_coord.points[1])-time_coord.units.num2date(time_coord.points[0])).seconds
    elif (time_spacing is not None):
        # use value of time_spacing for dt:
        dt=time_spacing
    return dxy,dt













```

**Comment by freemansw1** [comment]
@JuliaKukulies Are you still working with tobac? If so, I think we would be amenable to a pull request with this new function. Are we happy enough with the accuracy of the Haversine used here? I think it's probably fine and is especially fine given switching to the Vincenty or something else would probably require adding another dependency. If users want "the best" accuracy, they can calculate the projection themselves.

**Comment by JuliaKukulies** [comment]
@freemansw1 Perfect timing, I just started working on a vortex tracking project, where I am planning take up tobac again. Sure, I can work on a PR! As long as I can see, dxy is mainly used to calculate the distances in the feature detection and the propagation speed in the linking. So I guess for this purpose the haversine works just fine. Alternatively, we could allow the user to input a 2D field of grid deltas rather than one single value, e.g. if the user needs a better accuracy and has a grid with irregular spacing?

**Comment by freemansw1** [comment]
Great! Glad to hear it; I'm looking forward to the PR!

Yes, we could allow a 2D field of grid deltas, but I think that could end up being quite a bit of work and could, in some cases, be computationally expensive because when calculating the feature's position on the cartesian grid, we must now query its delta from the origin. Maybe I'm not thinking about this correctly, and there's a better/faster way to do that... I think eventually allowing users to track using a lat/lon grid rather than cartesian is the way to go, but we will need to replace the `dist_func` parameter when we call `trackpy` for something of our own. We (i.e., @galexsky and myself at CSU) are planning on replacing this function anyway for certain periodic boundary cases, so once that work is complete, hopefully, we'll have some insight on that process (and its computational efficiency).

**Comment by JuliaKukulies** [comment]
OK that sounds good! You are absolutely right. I had not thought so far yet, but the `dist_func` in `trackpy` seems to be the tricky (but probably not impossible) thing. We can give it some more thoughts and until then stay with a single value for `dxy`.

**Comment by freemansw1** [comment]
I think that this will be resolved (at least for what we have discussed here) in #65

**Comment by freemansw1** [comment]
Marking this as resolved given that #65 is merged.
